3.20.2 build error: `Error: posix_spawn(): Exec format error`

Could someone say what might be causing this failure with Dune 3.20.x? Dune 3.19.1 builds normally on the same system with everything else identical (these were successive builds, 3.19.1 worked and 3.20.2 upgrade failed).

<!-- Thank you for filing an issue to help us improve Dune! -->
## Expected Behavior
Build should succeed.

## Actual Behavior
```
--->  Configuring ocaml-dune
Executing:  cd "/opt/local/var/macports/build/ocaml-dune-3d7a7125/work/dune-3.20.2" && ./configure --bindir /opt/local/bin --datadir /opt/local/share/ocaml-dune --docdir /opt/local/share/doc --etcdir /opt/local/etc --libdir /opt/local/lib/ocaml/site-lib --mandir /opt/local/share/man --sbindir /opt/local/sbin 
--->  Building ocaml-dune
Executing:  cd "/opt/local/var/macports/build/ocaml-dune-3d7a7125/work/dune-3.20.2" && DUNE_CONFIG__COPY_FILE=portable /usr/bin/make release 
ocamlc -output-complete-exe -g -o .duneboot.exe -I boot unix.cma boot/[libs.ml](http://libs.ml/) boot/[duneboot.ml](http://duneboot.ml/)
./.duneboot.exe
Error: posix_spawn(): Exec format error
-> required by _build/default/doc/dune-cache.1
-> required by _build/install/default/man/man1/dune-cache.1
-> required by _build/default/dune.install
-> required by alias install
make: *** [release] Error 1
Command failed:  cd "/opt/local/var/macports/build/ocaml-dune-3d7a7125/work/dune-3.20.2" && DUNE_CONFIG__COPY_FILE=portable /usr/bin/make release 
Exit code: 2
Error: Failed to build ocaml-dune: command execution failed
Error: See /opt/local/var/macports/logs/_opt_local_ppcports_ocaml_ocaml-dune/ocaml-dune/main.log for details.
```

## Reproduction
Run the build.

## Specifications

- Version of `dune` (output of `dune --version`): 3.20.2
- Version of `ocaml` (output of `ocamlc --version`): 4.14.2
- Operating system (distribution and version): macOS 10.6.8

<!-- remove the following if not used -->
## Additional information 

MacPorts build env.

@rgrinberg Maybe you can suggest where to look?

--- a/otherlibs/stdune/src/io.ml
+++ b/otherlibs/stdune/src/io.ml
@@ -211,13 +211,23 @@ module Copyfile = struct
       copy_channels ic oc)
   ;;
 
-  let copy_file =
+  let copy_file_best =
     match available with
     | `Sendfile -> sendfile_with_fallback
     | `Copyfile -> copyfile
     | `Nothing -> copy_file_portable
   ;;
+
+  let copy_file_impl = ref `Best
+
+  let copy_file ?chmod ~src ~dst () =
+    match !copy_file_impl with
+    | `Portable -> copy_file_portable ?chmod ~src ~dst ()
+    | `Best -> copy_file_best ?chmod ~src ~dst ()
+  ;;
 end
+
+let set_copy_impl m = Copyfile.copy_file_impl := m
 
 module Make (Path : sig
     type t

--- a/otherlibs/stdune/src/io.mli
+++ b/otherlibs/stdune/src/io.mli
@@ -14,3 +14,5 @@ val portable_symlink : src:Path.t -> dst:Path.t -> unit
 
 (** Hardlink with fallback to copy on systems that don't support it. *)
 val portable_hardlink : src:Path.t -> dst:Path.t -> unit
+
+val set_copy_impl : [ `Portable | `Best ] -> unit

--- a/src/dune_config/config.ml
+++ b/src/dune_config/config.ml
@@ -102,6 +102,16 @@ let cutoffs_that_reduce_concurrency_in_watch_mode =
     ~default:`Disabled
 ;;
 
+let copy_file =
+  make
+    ~name:"copy_file"
+    ~of_string:(function
+      | "portable" -> Ok `Portable
+      | "fast" -> Ok `Best
+      | _ -> Error (sprintf "only %S and %S are allowed" "fast" "portable"))
+    ~default:`Best
+;;
+
 let background_default =
   match Platform.OS.value with
   | Linux | Windows | Darwin -> `Enabled

--- a/src/dune_config/config.mli
+++ b/src/dune_config/config.mli
@@ -39,6 +39,9 @@ val global_lock : Toggle.t t
     reduces concurrency *)
 val cutoffs_that_reduce_concurrency_in_watch_mode : Toggle.t t
 
+(** whether dune should optimize file copying on Linux/MacOS *)
+val copy_file : [ `Portable | `Best ] t
+
 (** Execute some actions in background threads. See [Action_exec] for the
     concrete list of actions *)
 val background_actions : Toggle.t t

--- a/src/dune_config_file/dune_config_file.ml
+++ b/src/dune_config_file/dune_config_file.ml
@@ -652,6 +652,7 @@ module Dune_config = struct
       | Preserve -> ()
       | Clear_on_rebuild -> Console.reset ()
       | Clear_on_rebuild_and_flush_history -> Console.reset_flush_history ());
+    Stdune.Io.set_copy_impl Config.(get copy_file);
     Log.verbose
     := match t.display with
        | Simple { verbosity = Verbose; _ } -> true

